### DISCUSSION TITLE
Promise 와 async/await 의 차이점

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#4] [프론트 기술 질문 공간] [answered]
Promise 와 async/await 의 차이점을 설명해주세요

**Comment by michoball** [comment, marked as answer]
동기적인 방식으로만 처리가 가능한 javascript 환경에서 비동기적인 처리를 가능하게 하기 위해서 es6 이후에 만들어진 문법입니다. 

 이전에는 함수를 비동기적으로 처리하기 위해서 callback을 사용했는데요. callback의 경우 중첩이 많이 된 함수를 처리함에 있어서 코드를 매우 복잡하게 만들고 가독성을 떨어트려 디버깅이 어렵다는 큰 단점이 있었습니다.  
 
이를 보완하고자 Promise를 만들었는데요. resolve, reject 두가지 형태로 값을 return하고 return 된 값에 대해서는  pending, rejected, fulfilled 의 3가지 상태를 가집니다. 

promise가 많은 부분 callback 함수의 단점을 보완했지만 promise로 resolve 또는 reject 된 값을 다루기 위해서는 then() 또는 catch() 와 같은 chaining이 필요하였고, 
이 promise chaining 또한 개발자에게  결국 불편함을 주어서 ECMAScript2017에 async await 이 등장하게 되었습니다. 

함수에 async를 붙여 함수를 비동기 함수로 만들고 그 안에서 비동기적으로 값을 처리해야하는 코드에 await을 붙여 비동기 처리가 끝날때까지 함수의 처리를 기다릴 수 있도록 했습니다. 

이로인해 promise chaining에서 벗어날 수 있었고, 코드 가독성을 올려주었습니다. 하지만 promise에 있는 promise all 이나 promise race 같이 async await 으로만 처리하지 못하는 부분이 완전히 promise를 대체하는 문법이라고 볼 수는 없습니다.

--- 

여기저기 찾아보면서 작성했는데 추가해야하거나 오류가 있으면 알려주세요~

**Comment by skyhanull** [comment]
promise는 기존의 callback함수의 콜백지옥을 보완하기 위해 나온 것입니다
프로미스는 3가지 상태를 가지는데 pending, fullfilled, rejected상태를 가집니다.
then을 이용한 체이닝으로 처리할 수 있고, 마지막에 catch 메서드를 통해 에러 핸들링을 할 수 있습니다.

위의 프로미스또한 체이닝을 하기에 불편함이 존재했기에 es6에서 async/await가 나오게됩니다.
함수앞에 async를 선언하면 promise 객체가 저절로 생성됩니다.
또한 async/await는 비동기 코드를 동기코드처럼 읽히게 해주어 가독성과 작성을 편하게 해줍니다
이때 async/await 는 try/catch문을 사용합니다.

**Comment by WooGie911** [comment]
JavaScript는 동기적이므로 비동기처리를 위해 callback 함수가 사용되었는데 그로인해 callback 지옥이라는 문제가 발생합니다.
이를 보완 하기 위해 Promise 라는 객체를 고안합니다. 
허나 then 과 catch를 사용하는 Promise를 체이닝 하는 과정에서 then 지옥이라는 문제가 발생하였고 
이를 해결하기 위해 async await 가 등장하게 되었습니다.
async await를 사용하면 가독성이 좋아지고 콜백,then지옥에서 벗어날수 있습니다.
허나 async await에는 에러 처리를 위한 catch 부분이 없기에 try catch 문과 함께 사용합니다.

**Comment by javascriptbiggosoo** [comment]
Promise 와 async/await 의 차이라는 말은 성립하지 않습니다.
async/await는 내부적으로 then/catch 방식 핸들링과 같은 로직이며 거기에 가독성을 부여한 문법적 설탕일 뿐입니다.

**Comment by BaekYuri** [comment]
두 문법 다 Javascript에서 비동기 처리를 하기 위해 사용하는 문법입니다.

Promise 
- Pending(대기), Fulfilled(이행), Rejected(실패) 세가지의 상태
- resolve를 실행하면 Fulfilled 상태, reject를 실행하면 Rejected 상태를 반환합니다.
- Fulfilled 상태는 then(), Reject 상태는 catch()을 통해 반환 받을 수 있습니다.

async/await
- function 앞에 async를 붙이면 promise를 반환합니다.
- promise 앞에 await를 붙이면 promise 객체가 아닌 일반 return 값이 반환됩니다.
- 에러 핸들링으로 try/catch 문을 활용합니다.

**Comment by bona373737** [comment]
자바스크립트에서 비동기적인 동작을 처리하기위한 전통적인 방식으로 콜백함수를 사용해 왔지만, 콜백의 깊이가 깊어지면 가독성이 크게 떨어지고, 비동기처리중 발생하는 에러의 처리가 까다로워지면서 그 해결책으로 promise객체를 활용한 비동기처리 방식이 등장하였다. 

Promise객체를 활용한 비동기처리방식은 Callback과 달리 비동기 처리 결과를 값으로 받아서 저장할 수 있다는 것이다. 반환받은 Promise 객체를 후속 처리 메서드(then, catch)를 통해 그 결과 값을 다루는 동작을 구현할 수 있다.  

async,await는 promise를 좀더 편리하게 사용할수 있게 등장한 문법으로, await은 async 함수 내의 Promise 앞에 쓰이면, async 함수의 실행을 일시중지하고 Promise 해결을 기다린 다음 함수의 실행을 재개시킨다.

**Comment by michoball** [comment]
async await 이 promise의 문법적인 불편함을 해소하게 해준 syntactic sugar 인 것에는 mdn에도 그렇게 쓰여있으니 이견은 없지만
promise.all 이나 promise.race 와 같은 기능은 async await으로는 표현할 수 없으니 둘에는 이부분에서 완전 동일한 기능을 한다고 보기 어렵지 않을까요? 다른 분들의 의견이 궁금합니다!
